packages/db/postgres/functions.sql

```diff
@@ -415,3 +416,37 @@ FROM cargo.aggregate_entries ae
 WHERE ae.cid_v1 = ANY (cids)
 ORDER BY de.entry_last_updated
 $$;
+
+
+-- a custom UPSERT operation for user account, so that we can distinguish between
+-- newly inserted users and updated ones.
+CREATE OR REPLACE FUNCTION upsert_user(_name TEXT, _picture TEXT, _email TEXT, _issuer TEXT, _github TEXT, _public_address TEXT)
+RETURNS TABLE (
+  "id" BIGINT,
+  "issuer" TEXT,
+  "inserted" BOOLEAN
+)
+LANGUAGE plpgsql
+AS
+$$
+#variable_conflict use_column
+DECLARE
+  inserted BOOLEAN;
+
+BEGIN
+  SELECT (COUNT(id) = 0) into inserted FROM public.user WHERE issuer = _issuer;
+
+  RETURN QUERY
+  INSERT INTO public.user AS u (name, picture, email, issuer, github, public_address) 
```

Comment:
probably not a big deal due to volume and how fast they would have to race:

but if called twice for the same user a milisecond or two apart won't it result in returning inserted = true for both?

Reply:
I think it's safe, since the `BEGIN` and `END` makes sure both statements happen in the same transaction.